### PR TITLE
Added `StringComparer` property to `TemplateContext`

### PR DESCRIPTION
**Pull request by gumbarros**
This PR allows case insensitivity at templates, making more easy to non-programmers to use variables and functions.

### CODE REVIEW

File: Fluid.Tests/TemplateContextTests.cs [hunk 1]
```diff
@@ -160,6 +160,15 @@ public async Task ShouldNotReleaseScopeAsynchronously()
...
             Assert.Equal("101", output);
         }
+        
+        [Fact]
+        public void ShouldUseStringComparer()
+        {
+            var context = new TemplateContext();
+            context.StringComparer = StringComparer.InvariantCultureIgnoreCase;
+            context.SetValue("PageState", "insert");
+            Assert.Equal("insert", context.GetValue("pageState").ToStringValue());
```
Comment (hishamco) [inline Fluid.Tests/TemplateContextTests.cs:170]:
```suggestion

            Assert.Equal("insert", context.GetValue("pageState").ToStringValue());
```

File: Fluid/Scope.cs [hunk 2]
```diff
@@ -7,28 +7,32 @@ public sealed class Scope
...
+            StringComparer = parent.StringComparer;
         }
 
         public Scope(Scope parent, bool forLoopScope)
         {
             if (forLoopScope && parent == null) ExceptionHelper.ThrowArgumentNullException(nameof(parent));
 
             Parent = parent;
-
+
```
Comment (hishamco) [inline Fluid/Scope.cs:29]:
```suggestion

```

File: Fluid/Scope.cs [hunk 3]
```diff
@@ -7,28 +7,32 @@ public sealed class Scope
...
-
+            
             // A ForLoop scope reads and writes its values in the parent scope.
             // Internal accessors to the inner properties grant access to the local properties.
             _forLoopScope = forLoopScope;
-
-            _properties = new Dictionary<string, FluidValue>();
+            StringComparer = parent.StringComparer;
+            _properties =
+                new Dictionary<string, FluidValue>(StringComparer);
```
Comment (hishamco) [inline Fluid/Scope.cs:35]:
```suggestion
            _properties = new Dictionary<string, FluidValue>(StringComparer);
```

File: Fluid/Scope.cs [hunk 4]
```diff
@@ -7,28 +7,32 @@ public sealed class Scope
     {
         private Dictionary<string, FluidValue> _properties;
         private readonly bool _forLoopScope;
+        public StringComparer StringComparer { get; set; }
```
Comment (hishamco) [inline Fluid/Scope.cs:10]: Move it after the constructors

**Review by hishamco — CHANGES_REQUESTED** [review]

File: Fluid/TemplateContext.cs [hunk 5]
```diff
@@ -150,6 +150,15 @@ public void IncrementSteps()
         /// </summary>
         public Func<string, string, ValueTask<string>> Captured { get; set; }
 
+        /// <summary>
+        /// Gets or sets the <see cref="StringComparer"/> used for matching keys of scope values.
+        /// </summary>
+        public StringComparer StringComparer
```
Comment (hishamco) [inline Fluid/TemplateContext.cs:156]: I think you can use the one that is declared in the `TemplateOptions`

https://github.com/sebastienros/fluid/blob/3f1750f690b5b2e02cd13e6554e0df832dbd8385/Fluid/TemplateContext.cs#L85

File: Fluid/TemplateContext.cs [hunk 6]
```diff
@@ -149,7 +149,7 @@ public void IncrementSteps()
         /// Gets or sets the delegate to execute when a Capture tag has been evaluated.
         /// </summary>
         public Func<string, string, ValueTask<string>> Captured { get; set; }
-
+
```
Comment (hishamco) [inline Fluid/TemplateContext.cs:152]:
```suggestion
```

File: Fluid/TemplateOptions.cs [hunk 7]
```diff
@@ -73,6 +73,14 @@ public class TemplateOptions
         /// </summary>
         public bool Greedy { get; set; } = true;
 
+        /// <summary>
+        /// Gets or sets the <see cref="StringComparer"/> used for matching keys of scope values.
+        /// </summary>
+        public StringComparer StringComparer
```
Comment (hishamco) [inline Fluid/TemplateOptions.cs:79]: Where this property is used?

**Review by hishamco — CHANGES_REQUESTED** [review]

File: Fluid/TemplateOptions.cs (hunk 7 above)
Comment (gumbarros) [reply Fluid/TemplateOptions.cs:79]: At my unit test to change the `StringComparer` all way down to the scope.

Comment (gumbarros) [reply Fluid/TemplateOptions.cs:79]: Wait, my test is failing because the property was removed.

Comment (hishamco) [reply Fluid/TemplateOptions.cs:79]: Can I fix this?

Comment (gumbarros) [reply Fluid/TemplateOptions.cs:79]: > Can I fix this?
Yes

Comment (hishamco) [reply Fluid/TemplateOptions.cs]: I will check ..

**Comment by hishamco** [comment]
@gumbarros check my last commit

**Comment by gumbarros** [comment]
> @gumbarros check my last commit

LGTM

File: Fluid/Scope.cs [hunk 8]
```diff
@@ -11,25 +11,31 @@ public sealed class Scope
         public Scope()
         {
             Parent = null;
+            StringComparer = StringComparer.CurrentCulture;
         }
```
Comment (hishamco) [inline Fluid/Scope.cs:15]:
```suggestion
        public Scope() : this(null)
        {
        }
```

**Review by hishamco — CHANGES_REQUESTED** [review]

**Review by hishamco — APPROVED** [review]

**Comment by hishamco** [comment]
@sebastienros anything else to add to this?

**Comment by hishamco** [comment]
Thanks for a quick fix :)

**Comment by gumbarros** [comment]
Any plans to merge this 👀?

**Comment by gumbarros** [comment]
I'm going to be out for 20 days next week and won't be able to make any changes for this PR. I would also like to use this feature this week if possible.

**Comment by hishamco** [comment]
@sebastienros any feedback on this or shall I merge?

**Comment by sebastienros** [comment]
This is too intrusive (in the ctor). Can this `StringComparer` be yet another property of the options/context classes, be inherited like the other properties?

And the new property should be named `ModelNamesComparer`? The default value should be Ordinal obviously.

**Comment by sebastienros** [comment]
@gumbarros 

if you can't do the changes @hishamco or myself should be able to do it.

**Comment by hishamco** [comment]
To be honest it was as it's seen in https://github.com/sebastienros/fluid/pull/681/commits/b99e0f54853fbebf42101758f24388917e2ac32f

If you like how it was I will revert the latest changes considering the name that you suggested

**Comment by sebastienros** [comment]
> To be honest it was as it's seen in https://github.com/sebastienros/fluid/commit/b99e0f54853fbebf42101758f24388917e2ac32f

Did I actually ask the opposite, or was it done on your own thoughts?

**Comment by hishamco** [comment]
It added as property at the beginning but I saw the property appears twice in different places that's why I suggested to pass the values through constructor

**Comment by sebastienros** [comment]
In you case that's what you mean, it's pretty common to see the properties duplicated and cloned from `TemplateOptions` and `TemplateContext`. But this is done this way for anything that can alter the behavior or the rendering, to have defaults or options to share across template content instances without having to set it each time.

**Comment by hishamco** [comment]
I see if the linked commit is fine for you I will revert the latest changes then merge unless you need to add something else

**Comment by gumbarros** [comment]
Hi guys, any news on merging this 👀?

**Comment by sebastienros** [comment]
Updated it. Should be simpler, and it helped me realize that this property can't be changed dynamically as it is used by an inner collection.

In Fluid 3.0 we might want to use the same comparer for both internal properties and model properties. (note for myself)

**Comment by nguyenngocanhpro** [comment]
When will this PR be released? I've been waiting for so long. 
And I think the `GetValueAsync` function from `FluidValue` should be updated in the same way as this case

**Comment by sebastienros** [comment]
@nguyenngocanhpro do you want to submit a PR for GetValueAsync? I can ship it right after that.

**Comment by nairdo** [comment]
With this PR (tested using v2.24.0), should we still be able to use **case-sensitive** variables if the `TemplateOptions ModelNamesComparer` is set to `StringComparer.Ordinal`?  

I was under the impression that Shopify Liquid still supports this behavior ([example](https://liquidjs.com/playground.html#IHslIGFzc2lnbiBjYXNlID0gImxvd2VyIiAlfSANCiB7JSBhc3NpZ24gQ0FTRSA9ICJ1cHBlciIgJX0gDQogY2FzZSA9IHt7IGNhc2UgfX0gPCEtLSBleHBlY3QgbG93ZXIgLS0+DQogQ0FTRSA9IHt7IENBU0UgfX0gPCEtLSBleHBlY3QgdXBwZXIgLS0+,ewogICJwZW9wbGUiOiBbCiAgICAiYWxpY2UiLAogICAgImJvYiIsCiAgICAiY2Fyb2wiCiAgXQp9Cg==)):

```Liquid
 {% assign case = "lower" %} 
 {% assign CASE = "upper" %} 
 case = {{ case }} <!-- expect lower -->
 CASE = {{ CASE }} <!-- expect upper -->
 ```
 
 Here's a unit test:
 ```CSharp
[Fact]
public void ShouldUseTemplateOptionsStringComparerWithCaseSensitive()
{
    var options = new TemplateOptions { ModelNamesComparer = StringComparer.Ordinal };
    var context = new TemplateContext(options);
    context.SetValue("case", "lower");
    context.SetValue("CASE", "upper");
    context.SetValue("Case", "mixed");

    Assert.Equal("lowerupper", context.GetValue("case").ToStringValue() + context.GetValue("CASE").ToStringValue());
}
 ```
 
 Results:

`     Actual:   "mixedmixed"`

**Comment by sebastienros** [comment]
The link you provided is from liquidjs, not shopify, here is the result from Shopify (https://shopify.dev/docs/api/liquid)

image

**Comment by nairdo** [comment]
@sebastienros Right, sorry about using liquidjs, I just mainly wanted to confirm that Fluid _should_ still allow case sensitivity.  I'll open this as a new issue.

**Comment by sebastienros** [comment]
I just wanted to be sure it was also the behavior in shopify. I also create issues on liquidjs because it doesn't follow the standard so I wanted to point it out.

**Comment by sebastienros** [comment]
Here is the PR fixing the bug: https://github.com/sebastienros/fluid/pull/802

**Pull request merged**